strawberryfields/plot.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+This module provides tools to visualize the state in various interactive
+ways using Plot.ly
+"""
+from copy import deepcopy
+
+import numpy as np
+
+plotly_error = (
+    "Plot.ly required for using this function. It can be installed using pip install "
+    "plotly or visiting https://plot.ly/python/getting-started/#installation"
+)
+
+try:
+    import plotly.io as pio
+except ImportError as e:
+    raise(plotly_error) from e
+
+
+def plot_wigner(state, p_axis, q_axis, renderer="browser", wire=0, dq=0.1):
+    """Plot the Wigner function with Plot.ly
+
+    Args:
+        state (:class:`.BaseState`): the state used for plotting
+        p_axis (int): Range of the discretized :math:`p` quadrature values.
+            E.g., ranging from ``-p_axis`` to ``p_axis``.
+        q_axis (int): Range of the discretized :math:`x` quadrature values.
+            E.g., ranging from ``-q_axis`` to ``q_axis``.
+        renderer (string): the renderer for plotting with Plot.ly
+        wire (int): wire used to calculate the reduced Wigner function
+        dq (float): stepsize along the p_axis and q_axis
+    """
+    xvec = np.arange(-q_axis, q_axis, dq)
+    pvec = np.arange(-p_axis, p_axis, dq)
```

Comment:
Just a thought (correct me if it's not valid): could it make sense to pre-define the landscape of `x` and `p` values in advance? Or make their values be optional

Reply:
It's a bit hard to pre-define the values without knowing (or at least, having a good guess) for the approximate domain of the state we are plotting 🤔 Maybe there is a rough heuristic between maximum photon number/cutoff dim and maximum quadrature value to plot?

@nquesada might know.

Reply:
Correct me if I'm wrong @josh146, but I think that the `[-cutoff, cutoff]` range was fairly arbitrarily chosen as a range that seemingly worked nicely. A better heuristic would be nice, but otherwise I feel like this could be left to the user to decide, and for the simulators we could just pass the cutoff values as ranges if we cannot think of a better way.

The other issue is that only the Fock backend would have a cutoff. So it would only work in that specific case then. In SF interactive we always pass a cutoff value, even when using another backend.

Reply:
Yep, pretty much @thisac. I chose the `[-cutoff, cutoff]` domain very arbitrarily ~ 2 years ago. One downside is that for states with a large cutoff, the plotting domain is huge even if the state to be plotted is very close to the origin!

@antalszava I think this comment thread is now outdated -- the latest version of this PR (as far as I can tell) has the plotting function accept the domain directly as an argument.

Reply:
I think it does not make much sense to allow for predefined ranges in phase space. It is certainly true that a state with `cutoff` Fock space support will be localized to a circle of radius of sqrt(cutoff) in phase space (Josh I think was being too generous two years ago). Yet, I think the burden of picking ranges should be passed to the user, deciding this in advance is just too tricky.

Reply:
>Josh I think was being too generous two years ago

😆